### PR TITLE
Memory

### PR DESCRIPTION
**Pull request by piotrva**

### CODE REVIEW

File: Firmware/core/storage.cpp [hunk 1]
```diff
@@ -8,40 +8,51 @@
...
 
 void read() {
-    HAL::Memory::get(inMemory.asArrayView());
-    core::target = inMemory.targetTemperature;
-    core::pid.params = inMemory.pidParams;
-    core::tempSensor::params = inMemory.tipParams;
-    core::display::setContrast(inMemory.contrast);
-    core::display::setBacklight(inMemory.backlight);
+    HAL::Memory::getSettings(settingsInMemory.asArrayView());
+    HAL::Memory::getState(stateInMemory.asArrayView());
```
Comment (ggajoch) [inline Firmware/core/storage.cpp]: extract 'state' usage to line 33 and below (make it visually separate from settings)

File: Firmware/core/storage.cpp [hunk 2]
```diff
@@ -2,46 +2,76 @@
...
+        static constexpr core::PID::Params pidParams = {.Kp = 0.0, .Ki = 0.0, .Kd = 0.0};
+        static constexpr core::tempSensor::Params tipParams = {.offset = 0, .gain = 0.0};
+        core::pid.params = pidParams;
+        core::tempSensor::params = tipParams;
+        core::display::setContrast(27.5);
+        core::display::setBacklight(100.0);
+    } else {
+        core::sleepManager::configStateSet(true);
+        settingsInMemory = *settings;
+        core::pid.params = settingsInMemory.pidParams;
```
Comment (ggajoch) [inline Firmware/core/storage.cpp:39]: refactor to prevent code duplication

File: Firmware/core/display.cpp [hunk 3]
```diff
@@ -13,7 +13,9 @@ void tick() {
 
     std::sprintf(display[0], "%3d/%3d     %3d%%", static_cast<int>(temp), static_cast<int>(target), static_cast<int>(power));
 
-    if (sleepManager::sleepState) {
+    if (!sleepManager::configState) {
```
Comment (ggajoch) [inline Firmware/core/display.cpp]: why NOT configState?

File: Firmware/core/sleepManager.h [hunk 4]
```diff
@@ -4,10 +4,11 @@
 namespace core {
 namespace sleepManager {
 
-extern bool sleepState, standState;
+extern bool sleepState, standState, configState;
```
Comment (ggajoch) [inline Firmware/core/sleepManager.h:7]: is it needed?

File: Firmware/targets/arm/Inc/hardwareConfig.h [hunk 5]
```diff
@@ -4,5 +4,6 @@
 constexpr static int ENCODER_DEBOUNCE_STABLE = 5;
 constexpr static int BUTTON_DEBOUNCE_STABLE = 20;
 constexpr static int ENCODER_PRESC = 4;
+constexpr enum{MEM_24C16, MEM_24C08, MEM_24C04, MEM_24C02, MEM_24C01} MEMORY_TYPE = MEM_24C16;
```
Comment (ggajoch) [inline Firmware/targets/arm/Inc/hardwareConfig.h:7]: extract enum to memory driver, enum definition shouldn't be in config file

File: Firmware/targets/arm/Src/HAL.cpp [hunk 6]
```diff
@@ -108,18 +110,45 @@ void callbackTick() {
...
+std::experimental::optional<core::storage::Settings> getSettings() {
+    core::storage::Settings elements;
+
+    auto fromMemory = i2cMemoryReadSettings();
+
+    if (!fromMemory) {
+        return {};
+    }
+
+    i2cMemorySettings formMemorySettings = *fromMemory;
```
Comment (ggajoch) [inline Firmware/targets/arm/Src/HAL.cpp]: extract this to some MemoryLayout module or sth like this

Comment (piotrva) [reply Firmware/targets/arm/Src/HAL.cpp]: Can you explain it clearer?

File: Firmware/core/display.cpp (hunk 3 above)
Comment (ggajoch) [inline Firmware/core/display.cpp:21]: why negation?

**Pull request merged**